sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdServiceEndpoint.java

```diff
@@ -377,9 +377,9 @@ private RntbdRequestRecord writeWhenConnected(
         if (connected.isSuccess()) {
             final Channel channel = (Channel) connected.getNow();
             assert channel != null : "impossible";
-            this.releaseToPool(channel);
             requestRecord.channelTaskQueueLength(RntbdUtils.tryGetExecutorTaskQueueSize(channel.eventLoop()));
-            channel.write(requestRecord.stage(RntbdRequestRecord.Stage.PIPELINED));
+            channel.write(requestRecord.stage(RntbdRequestRecord.Stage.PIPELINED))
+                .addListener(ignore -> this.releaseToPool(channel));
```

Comment:
why do we need to change the order of `releaseToPool(.)` invocation?

Reply:
The main purpose of releaseToPool is to check the healthy of the channel, and then add the channel to availableChannels so it can be reused by other requests.  Changing the order due to the following reasons:
1. There is a race condition that we might pipeline > maxRequestsPerChannel to the channel. 
2. Release channel also need to happen on the IO thread, after the request acquire channel, it feels more right to first write as soon as possible then release

Reply:
I don't see a correctness issue in the existing code.

However my worry is with this change of ordering we may see perf degradation for for high QPS scenario.

With this new approach, the channel cannot be used for a longer duration. this means if you have high QPS scenario. This channel can't be used for a while -> perf degradation

However I still don't see why this can help to improve latency for low QPS scenario that we are investigating? I don't think this would help for improving low QPS scenario latency (it may help for reducing latency for high QPS scenario as you explained in your 1. but doesn't improve for low QPS scenario. ?

Reply:
This change shows perf regression for low QPS. As discussed offline, reverted the change back.